test/integration/hooks_spec.js

```diff
@@ -0,0 +1,208 @@
+                    });
+            });
+        },
+
+        destroy: function (data) {
+            return bookshelf.transaction((transacting) => {
+                return this.forge(_.pick(data, 'id'))
+                    .destroy({transacting: transacting});
+            });
+        }
+    });
+
+    result.Tag = bookshelf.model('Tag', Tag);
+    result.Post = bookshelf.model('Post', Post);
+    return result;
+}
+
+describe('[Integration] Hooks: Posts/Tags', function () {
+    let authorId;
+    beforeEach(function () {
+        return testUtils.database.reset()
+            .then(function () {
+                return testUtils.database.init();
+            })
+            .then(function () {
+                const knex = testUtils.database.getConnection();
+                return knex('authors').select('id').first().then((row) => {
+                    authorId = row.id;
+                });
+            });
+    });
+
+    it('Uses the model hooks if all hooks present', function () {
+        const result = setupModel(testUtils.database.getConnection(), {
+            global: {
+                before: true,
+                after: true
+            },
+            model: {
+                before: true,
+                after: true
+            }
+        });
+
+        return result.Post.add({
+            author_id: authorId,
+            tags: ['blah']
+        }).then(() => {
+            should.equal(result.globalAfterHookCalled, false);
+            should.equal(result.globalBeforeHookCalled, false);
+            should.equal(result.modelAfterHookCalled, true);
+            should.equal(result.modelBeforeHookCalled, true);
+        });
+    });
+
+    it('Uses the global hooks if no model hook there', function () {
+        const result = setupModel(testUtils.database.getConnection(), {
+            global: {
+                before: true,
+                after: true
+            },
+            model: {
+                before: false,
+                after: false
+            }
+        });
+
+        return result.Post.add({
+            author_id: authorId,
+            tags: ['blah']
+        }).then(() => {
+            should.equal(result.globalAfterHookCalled, true);
+            should.equal(result.globalBeforeHookCalled, true);
+            should.equal(result.modelAfterHookCalled, false);
+            should.equal(result.modelBeforeHookCalled, false);
+        });
+    });
+
+    it('Uses the model hooks when available', function () {
+        const result = setupModel(testUtils.database.getConnection(), {
+            global: {
+                before: true,
+                after: true
+            },
+            model: {
+                before: false,
+                after: true
+            }
+        });
+
+        return result.Post.add({
+            author_id: authorId,
+            tags: ['blah']
+        }).then(() => {
+            should.equal(result.globalAfterHookCalled, false);
```

Comment:
What was your reasoning behind overriding the global hook completely?

I'm not sure about this behavior tbh. Usually when there are global hooks I'd expect them to "stack up" on to of each other. For example, the model hook would be called if there's one defined but than the global hook would be callled after. Can imagine there being some global behavior - let's set log some info, and then some model-specific behavior. From developer perspective I'd expecte all of them to be executed if all are defined: modelBeforeHook -> globalBeforeHook -> modelAfterHook -> globalAfterHook 🤔

Reply:
My reasoning was that if you have a global hook which does some operation, you'd have no way to stop the operation in the model hook. This feels much simpler to me and easier to reason about

Reply:
If we want to stop the global operation we should be calling something more explicit in the model layer hook then. Overriding the global hook without a warning or explicit logic might lead to more trouble imo

Reply:
We discussed this privately and came to the conclusion that without concrete usecases there's no need to get super picky about this behavior. There's also not straight forward reference to look at in other projects.